src/listening_groups.py

```diff
@@ -0,0 +1,68 @@
+import csv
+import os.path
+
+from core_data_modules.logging import Logger
+from core_data_modules.traced_data import Metadata
+import time
+
+from src.lib import PipelineConfiguration
+
+log = Logger(__name__)
+
+
+class ListeningGroups(object):
+    @classmethod
+    def tag_listening_groups_participants(cls, user, data, pipeline_configuration, listening_group_dir):
+        """
+        This tags uids who participated in repeat listening groups and/or weekly listening
+        group sessions.
+        :param user: Identifier of the user running this program, for TracedData Metadata.
+        :type user: str
+        :param data: TracedData objects to tag listening group participation to.
+        :type data: iterable of TracedData
+        :param listening_group_dir: Directory containing de-identified listening groups contacts CSVs containing
+                                    listening groups data stored as `Name` and `avf-phone-uuid` columns.
+        :type user: str
+        :param pipeline_configuration: Pipeline configuration.
+        :type pipeline_configuration: PipelineConfiguration
+        """
+        repeat_listening_group_participants = [] # Contains uids of listening group participants who will participate
+                                                 # in all listening group sessions.
+        listening_group_participants = dict()   # Contains lists of weekly listening group participants. The participants
+                                                # will change each week.
+
+        # Read repeat listening group participants CSV and add their uids to repeat_listening_group_participants lists
+        with open(f'{listening_group_dir}/repeat_listening_group.csv', "r", encoding='utf-8-sig') as f:
+            repeat_listening_group_data = list(csv.DictReader(f))
+            for row in repeat_listening_group_data:
+                repeat_listening_group_participants.append(row['avf-phone-uuid'])
+            log.info(f'Loaded {len(repeat_listening_group_participants)} repeat listening group participants')
+
+        # Read weekly listening group participants CSVs and add their uids to the respective radio-show
+        # listening_group_participants lists
+        listening_group_csvs = []
+        for listening_group_csv_url in pipeline_configuration.listening_group_csv_urls:
+            listening_group_csvs.append(listening_group_csv_url.split("/")[-1])
+        for plan in PipelineConfiguration.RQA_CODING_PLANS:
+            listening_group_participants[plan.dataset_name] = []
+            if f'{plan.dataset_name}_listening_group.csv' in listening_group_csvs:
```

Comment:
I'm not convinced there's no longer a risk of data silently not being imported, because this requires the listening groups in configuration to be in exactly this naming format.

It would have been even cleaner to add this file name to the CodingPlans, because then you no longer require these to be in a specific format, and that would be consistent with the other dataset-specific files we read/write. Alternatively, you could add a line to the pipeline configuration file parser that asserts that all listening groups are in this format.

Reply:
The files are provided on a weekly/bi-weekly basis so not available from the first pipeline run. We, therefore, have to check whether the file is available before loading it, otherwise, the pipeline will crash. Although I have added the file name to the CodingPlans, it does not prevent the risk of the file not being silently imported. We have little control over the file naming as the file is created outside the pipeline. @as2388 PTAL!